docs/orleans/grains/grain-references.md

```diff
@@ -0,0 +1,214 @@
+
+> [!NOTE]
+> Orleans generates grain reference implementation types for each grain interface in your application  during compilation. These grain reference implementations inherit from the <xref:Orleans.Runtime.GrainReference?displayProperty=nameWithType> class. `GetGrain` returns instances of the generated <xref:Orleans.Runtime.GrainReference?displayProperty=nameWithType> implementation corresponding to the requested grain interface.
+
+## Disambiguating grain type resolution
+
+When there are multiple implementations of a grain interface, such as in the following example, Orleans attempts to determine the intended implementation when creating a grain reference. Consider the following example, in which there are two implementations of the `ICounterGrain` interface:
+
+```csharp
+public interface ICounterGrain : IGrainWithStringKey
+{
+  ValueTask<int> UpdateCount();
+}
+
+public class UpCounterGrain : ICounterGrain
+{
+  private int _count;
+  public ValueTask<string> UpdateCount() => new(++_count); // Increment count
+}
+
+public class DownCounterGrain : ICounterGrain
+{
+  private int _count;
+  public ValueTask<string> UpdateCount() => new(--_count); // Decrement count
+}
+```
+
+The following call to `GetGrain` will throw an exception because Orleans does not know how to unambiguously map `ICounterGrain` to one of the grain classes.
+
+```csharp
+// This will throw an exception: there is no unambiguous mapping from ICounterGrain to a grain class.
+ICounterGrain myCounter = grainFactory.GetGrain<ICounterGrain>("my-counter");
+```
+
+An <xref:System.ArgumentException?displayProperty=nameWithType> will thrown with the following message:
+
+```
+Unable to identify a single appropriate grain type for interface ICounterGrain. Candidates: upcounter (UpCounterGrain), downcounter (DownCounterGrain)
+```
+
+The error message tells us which grain implementations Orleans has which match the requested grain interface type, `ICounterGrain`. It shows us the grain type names (`upcounter` and `downcounter`) as well as the grain classes (`UpCounterGrain` and `DownCounterGrain`).
+
+> [!NOTE] The grain type names in the above error message, `upcounter` and `downcounter`, are derived from the grain class names, `UpCounterGrain` and `DownCounterGrain` respectively. This is the default behavior in Orleans and it can be customized by adding a `[GrainType(string)]` attribute to the grain class. For example:
```

Comment:
```suggestion
> [!NOTE]
> The grain type names in the preceding error message, `upcounter` and `downcounter`, are derived from the grain class names, `UpCounterGrain` and `DownCounterGrain` respectively. This is the default behavior in Orleans and it can be customized by adding a `[GrainType(string)]` attribute to the grain class. For example:
```